crash in raft.sendAppend: error from range_raftstorage.go:245

```
E0528 20:47:12.818151       1 queue.go:262] failure processing range range=1 (""-"\xff\xff") from replicate queue: allocator.go:207: unable to allocate a target store for {[] }
I0528 20:47:12.849569       1 client.go:150] client roach1.local:8080 connected
I0528 20:47:12.849768       1 client.go:224] client roach1.local:8080 starting heartbeat
E0528 20:47:13.210226       1 queue.go:262] failure processing range range=1 (""-"\xff\xff") from replicate queue: allocator.go:207: unable to allocate a target store for {[] }
E0528 20:47:13.622575       1 queue.go:262] failure processing range range=1 (""-"\xff\xff") from replicate queue: allocator.go:207: unable to allocate a target store for {[] }
E0528 20:47:14.024363       1 queue.go:262] failure processing range range=1 (""-"\xff\xff") from replicate queue: allocator.go:207: unable to allocate a target store for {[] }
E0528 20:47:14.428034       1 queue.go:262] failure processing range range=1 (""-"\xff\xff") from replicate queue: allocator.go:207: unable to allocate a target store for {[] }
I0528 20:47:14.563834       1 server.go:111] gossip: received 7 info(s) from 172.17.0.191:8080
I0528 20:47:14.901081       1 client.go:150] client roach2.local:8080 connected
I0528 20:47:14.901209       1 client.go:224] client roach2.local:8080 starting heartbeat
panic: range_raftstorage.go:245: failed to get desc: conflicting write intents [{Key:"\x00\x00\x00k\x00\x01rdsc" Txn:"change replicas of 1" {id=d751d638-91ea-41f4-88a2-e946d05ebe02 pri=1579927167, iso=SERIALIZABLE, stat=PENDING, epo=0, ts=1432846034.852988720,0 orig=1432846034.852988720,0 max=1432846035.102988482,0} XXX_unrecognized:[]}]: resolved? false

goroutine 19 [running]:
github.com/coreos/etcd/raft.(*raft).sendAppend(0xc208182000, 0x300000003)
	/go/src/github.com/coreos/etcd/raft/raft.go:250 +0x159
github.com/coreos/etcd/raft.stepLeader(0xc208182000, 0x4, 0x100000001, 0x300000003, 0x7, 0x0, 0xa3, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/coreos/etcd/raft/raft.go:523 +0x6f3
github.com/coreos/etcd/raft.(*raft).Step(0xc208182000, 0x4, 0x100000001, 0x300000003, 0x7, 0x0, 0xa3, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/coreos/etcd/raft/raft.go:487 +0x259
github.com/coreos/etcd/raft.(*multiNode).run(0xc208112840)
	/go/src/github.com/coreos/etcd/raft/multinode.go:236 +0xd0b
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:56 +0xb6
```
full log: https://gist.github.com/tamird/b4cdde131fe7aa464e92

discovered when running PutTest with 5 nodes.

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -18,6 +18,7 @@
 package storage_test
 
 import (
+	"fmt"
 	"reflect"
 	"strings"
 	"sync"
@@ -728,5 +729,45 @@ func TestReplicateAfterSplit(t *testing.T) {
 		return getResp.Value.GetInteger() == 11
 	}, 1*time.Second); err != nil {
 		t.Fatal(err)
+	}
+}
+
+// TestRangeDescriptorSnapshotRace calls Snapshot() repeatedly while
+// transactions are performed on the range descriptor.
+func TestRangeDescriptorSnapshotRace(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	mtc := startMultiTestContext(t, 1)
+	defer mtc.Stop()
+
+	rng, err := mtc.stores[0].GetRange(1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	stopper := util.NewStopper()
+	defer stopper.Stop()
+	// Call Snapshot() in a loop and ensure it never fails.
+	stopper.RunWorker(func() {
+		for {
+			select {
+			case <-stopper.ShouldStop():
+				return
+			default:
+				_, err := rng.Snapshot()
+				if err != nil {
+					t.Fatal(err)
+				}
+			}
+		}
+	})
+
+	// Split the range 100 times (carving chunks off the end of the initial range).
+	for i := 100; i > 0; i-- {
+		args, reply := adminSplitArgs(proto.KeyMin, []byte(fmt.Sprintf("%03d", i)), 1, mtc.stores[0].StoreID())
+		rng.AdminSplit(args, reply)
+		if reply.GoError() != nil {
+			t.Fatal(reply.GoError())
+		}
 	}
 }

--- a/storage/engine/mvcc.go
+++ b/storage/engine/mvcc.go
@@ -312,21 +312,24 @@ func MVCCSetRangeStats(engine Engine, raftID proto.RaftID, ms *proto.MVCCStats) 
 
 // MVCCGetProto fetches the value at the specified key and unmarshals
 // it using a protobuf decoder. Returns true on success or false if
-// the key was not found.
+// the key was not found. In the event of a WriteIntentError when
+// consistent=false, we return the error and the decoded result; for
+// all other errors (or when consistent=true) the decoded value is
+// invalid.
 func MVCCGetProto(engine Engine, key proto.Key, timestamp proto.Timestamp, consistent bool, txn *proto.Transaction, msg gogoproto.Message) (bool, error) {
 	value, err := MVCCGet(engine, key, timestamp, consistent, txn)
-	if err != nil {
-		return false, err
-	}
-	if value == nil || len(value.Bytes) == 0 {
-		return false, nil
-	}
-	if msg != nil {
-		if err := gogoproto.Unmarshal(value.Bytes, msg); err != nil {
-			return true, err
+	found := value != nil && len(value.Bytes) > 0
+	// If we found a result, parse it regardless of the error returned by MVCCGet.
+	if found && msg != nil {
+		unmarshalErr := gogoproto.Unmarshal(value.Bytes, msg)
+		// If the unmarshal failed, return its result. Otherwise, pass
+		// through the underlying error (which may be a WriteIntentError
+		// to be handled specially alongside the returned value).
+		if unmarshalErr != nil {
+			err = unmarshalErr
 		}
 	}
-	return true, nil
+	return found, err
 }
 
 // MVCCPutProto sets the given key to the protobuf-serialized byte
@@ -369,9 +372,9 @@ var getBufferPool = sync.Pool{
 // keyB : MVCCMetadata of keyB
 // ...
 //
-// The consistent parameter indicates that intents should cause
-// WriteIntentErrors. If set to false, intents are ignored; keys with
-// an intent but no earlier committed versions, will be skipped.
+// The consistent parameter indicates how write intents should be
+// handled. In inconsistent mode, we return a valid value if one was
+// found in addition to the WriteIntentError.
 func MVCCGet(engine Engine, key proto.Key, timestamp proto.Timestamp, consistent bool, txn *proto.Transaction) (*proto.Value, error) {
 	if len(key) == 0 {
 		return nil, emptyKeyError()

--- a/storage/engine/mvcc_test.go
+++ b/storage/engine/mvcc_test.go
@@ -688,6 +688,93 @@ func TestMVCCGetInconsistent(t *testing.T) {
 	}
 }
 
+// TestMVCCGetProtoInconsistent verifies the behavior of GetProto with
+// consistent set to false.
+func TestMVCCGetProtoInconsistent(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	engine := createTestEngine()
+	defer engine.Close()
+
+	value1, err := gogoproto.Marshal(&proto.RawKeyValue{Value: []byte("value1")})
+	if err != nil {
+		t.Fatal(err)
+	}
+	value2, err := gogoproto.Marshal(&proto.RawKeyValue{Value: []byte("value2")})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Put two values to key 1, the latest with a txn.
+	if err := MVCCPut(engine, nil, testKey1, makeTS(1, 0), proto.Value{Bytes: value1}, nil); err != nil {
+		t.Fatal(err)
+	}
+	if err := MVCCPut(engine, nil, testKey1, makeTS(2, 0), proto.Value{Bytes: value2}, txn1); err != nil {
+		t.Fatal(err)
+	}
+
+	// A get with consistent=false should fail in a txn.
+	val := &proto.RawKeyValue{}
+	_, err = MVCCGetProto(engine, testKey1, makeTS(1, 0), false, txn1, val)
+	if err == nil {
+		t.Error("expected an error getting with consistent=false in txn")
+	} else if _, ok := err.(*proto.WriteIntentError); ok {
+		t.Error("expected non-WriteIntentError with inconsistent read in txn")
+	}
+
+	// Inconsistent get will fetch value1 for any timestamp.
+	for _, ts := range []proto.Timestamp{makeTS(1, 0), makeTS(2, 0)} {
+		val.Reset()
+		found, err := MVCCGetProto(engine, testKey1, ts, false, nil, val)
+		if ts.Less(makeTS(2, 0)) {
+			if err != nil {
+				t.Fatal(err)
+			}
+		} else {
+			if wiErr, ok := err.(*proto.WriteIntentError); !ok || !wiErr.Intents[0].Key.Equal(testKey1) {
+				t.Fatal(err)
+			}
+		}
+		if !found {
+			t.Errorf("expected to find result with inconsistent read")
+		}
+		if !bytes.Equal(val.Value, []byte("value1")) {
+			t.Errorf("@%s expected %q; got %q", ts, []byte("value1"), val.Value)
+		}
+	}
+
+	// Write a single intent for key 2 and verify get returns empty.
+	if err := MVCCPut(engine, nil, testKey2, makeTS(2, 0), proto.Value{Bytes: value1}, txn2); err != nil {
+		t.Fatal(err)
+	}
+	found, err := MVCCGetProto(engine, testKey2, makeTS(2, 0), false, nil, val)
+	if wiErr, ok := err.(*proto.WriteIntentError); !ok || !wiErr.Intents[0].Key.Equal(testKey2) {
+		t.Fatal(err)
+	}
+	if found {
+		t.Errorf("expected no result; got %+v", val)
+	}
+
+	// Write a malformed value (not an encoded proto.RawKeyValue) and a
+	// write intent to key 3; the parse error is returned instead of the
+	// write intent.
+	if err := MVCCPut(engine, nil, testKey3, makeTS(1, 0), value3, nil); err != nil {
+		t.Fatal(err)
+	}
+	if err := MVCCPut(engine, nil, testKey3, makeTS(2, 0), proto.Value{Bytes: value2}, txn1); err != nil {
+		t.Fatal(err)
+	}
+	val.Reset()
+	found, err = MVCCGetProto(engine, testKey3, makeTS(1, 0), false, nil, val)
+	if err == nil {
+		t.Errorf("expected error reading malformed data")
+	} else if _, ok := err.(*proto.WriteIntentError); ok {
+		t.Errorf("expected non-WriteIntentError with malformed data")
+	}
+	if !found {
+		t.Errorf("expected to find result with malformed data")
+	}
+}
+
 func TestMVCCScan(t *testing.T) {
 	defer leaktest.AfterTest(t)
 	engine := createTestEngine()

--- a/storage/range_raftstorage.go
+++ b/storage/range_raftstorage.go
@@ -242,8 +242,13 @@ func (r *Range) Snapshot() (raftpb.Snapshot, error) {
 	ok, err := engine.MVCCGetProto(snap, keys.RangeDescriptorKey(r.Desc().StartKey),
 		r.rm.Clock().Now(), false, nil, &desc)
 	if err != nil {
-		return raftpb.Snapshot{}, util.Errorf("failed to get desc: %s", err)
-	} else if !ok {
+		// MVCCGetProto may return WriteIntentErrors in addition to a valid value.
+		// We can't resolve intents at this level so just ignore the error.
+		if _, ok := err.(*proto.WriteIntentError); !ok {
+			return raftpb.Snapshot{}, util.Errorf("failed to get desc: %s", err)
+		}
+	}
+	if !ok {
 		return raftpb.Snapshot{}, util.Errorf("couldn't find range descriptor")
 	}
 

